### PR TITLE
Add definitions

### PR DESCRIPTION
**Pull request by joaohf**
Basically definitions objects holds data types that are consumed and
produced by operations.

This allows definitions to be defined using metadata like this:

    Metadata = #{
       get ....
       post ....
       definitions => #{
           vendor => #{
              type => <<"object">>,
              required => [ <<"id">>, <<"name">>],
              properties => #{
                <<"id">> => #{
                    type => <<"integer">>,
                    format => <<"int64">>
                   },
                <<"name">> => #{
                    type => <<"string">>
                   }
               }
             }
        }}

So get, post can use to references schema like this:

    parameters => [
        #{name => <<"vendor">>,
          in => <<"body">>,
          description => <<"Vendor to add">>,
          required => true,
          schema => #{
              <<"$ref">> => <<"#/definitions/newVendor">>
          }
        }
    ],
    responses => #{
                  <<"200">> => #{description => <<"Succesful operation">>,
                                 schema => #{<<"$ref">> => <<"#/definitions/vendor">>}}

### CODE REVIEW

File: src/cowboy_swagger.erl [hunk 1]
```diff
@@ -120,6 +131,13 @@ normalize_path(Path) ->
     "\\[|\\]|\\:", "", [{return, binary}, global]).
 
 %% @private
+swagger_definitions([], Acc) ->
+  Acc;
+swagger_definitions([Trail | T], Acc) ->
+  Definitions = normalize_map_values(validate_metadata_definitions(trails:metadata(Trail))),
```
Comment (gadgetci) [inline src/cowboy_swagger.erl]: According to **Elvis**:
> Line 137 is too long:   Definitions = normalize_map_values(validate_metadata_definitions(trails:metadata(Trail))),.

File: src/cowboy_swagger.erl [hunk 2]
```diff
@@ -110,7 +121,7 @@ swagger_paths([], Acc) ->
   Acc;
 swagger_paths([Trail | T], Acc) ->
   Path = normalize_path(trails:path_match(Trail)),
-  Metadata = normalize_map_values(validate_metadata(trails:metadata(Trail))),
+  Metadata = normalize_map_values(validate_metadata_path(trails:metadata(Trail))),
```
Comment (gadgetci) [inline src/cowboy_swagger.erl]: According to **Elvis**:
> Line 124 is too long:   Metadata = normalize_map_values(validate_metadata_path(trails:metadata(Trail))),.

**Comment by joaohf** [comment]
Hi,

This PR is the general idea. Am I in the right direction?

**Comment by elbrujohalcon** [comment]
Hi @joaohf. I wouldn't tie those definitions to a single handler.
In my mind they should be somewhat global.
Therefore, they should be in the application environment. Something like

```erlang
Definitions = application:get_env(cowboy_swagger, definitions, [])
```

Which is what we're doing today with `global_spec`. As a matter of fact, can't they be included there already?

**Comment by joaohf** [comment]
Hi,

Ok, Makes sense.

So will use application environment.

Thanks.

**Pull request closed without merging**